pandas/core/util/_dispatching.py

```diff
@@ -0,0 +1,186 @@
+import collections
+import inspect
```

Comment:
rather than put this here and trying to make this too generic. let's do:

- move ``pandas.core.groupby`` to ``pandas.core.groupby.groupby`` (and add a shim to deprecate ``pandas.core.groupby``, simliar to what we do for ``pandas.core.categorical``), do as a pre-cursor PR, only doing this move
- put this module in ``pandas.core.groupby``, call it ``dispatch.py``